lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp

```diff
@@ -418,6 +418,116 @@ class RecomposeChunkListUnpack : public OpRewritePattern<PrimListUnpackOp> {
     return success();
   }
 };
+
+// recompose `aten::nonzero_numpy` + `prim::ListUnpack` + `prim::ListConstruct` + `aten::index.Tensor` with
+// `aten::masked_select`
+class RecomposeIndexTensorWithNonzeroNumpy
+    : public OpRewritePattern<AtenIndexTensorOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  // Look forward a few steps in order to try to restore the shape information
+  // of the value. For now, it only handles with static_info_cast and
+  // copy.to_tensor op.
+  static FailureOr<Value> getOriginalValue(Value input) {
+    auto inputType = input.getType().cast<BaseTensorType>();
+    while (!inputType.hasSizes()) {
+      if (auto staticInfoCastOp =
+              input.getDefiningOp<TensorStaticInfoCastOp>()) {
+        input = staticInfoCastOp.getOperand();
+        inputType = input.getType().cast<BaseTensorType>();
+      } else if (auto copyToNonValueTensorOp =
+                     input.getDefiningOp<CopyToNonValueTensorOp>()) {
+        input = copyToNonValueTensorOp.getOperand();
+        inputType = input.getType().cast<BaseTensorType>();
+      } else {
+        return failure();
+      }
+    }
+    return input;
+  }
+  LogicalResult matchAndRewrite(AtenIndexTensorOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    MLIRContext *context = op.getContext();
+
+    Value input = op.getSelf();
+    Value index = op.getIndices();
+    AtenNonzeroNumpyOp nonzeroNumpyOp =
+        index.getDefiningOp<AtenNonzeroNumpyOp>();
+    if (!nonzeroNumpyOp) {
+      return rewriter.notifyMatchFailure(op,
+                                         "index of aten::index.tensor should "
+                                         "come from aten::nonzero_numpy op");
+    }
+
+    Value mask = nonzeroNumpyOp.getSelf();
+    auto inputType = input.getType().cast<BaseTensorType>();
+    auto maskType = mask.getType().cast<BaseTensorType>();
+
+    // Because this pass is run before reduce-op-variants and
+    // maximize-value-semantics pass, operation operands might already be
+    // converted through static_info_cast or copy.to_tensor op. We need to
+    // restore the shape information to judge whether we can convert the op.
+    auto originalMaskInfo = getOriginalValue(mask);
+    auto originalInputInfo = getOriginalValue(input);
```

Comment:
I'm a bit confused by the need for `getOriginalValue`. Can you show me the IR you're seeing that causes issues? I would expect that returning `failure()` when the tensors don't have shape information would work, since the pipeline is applied multiple times, so at some point the shape information would propagate all the way.